### DISCUSSION TITLE
#Review Affiction/angular-feature-sliced

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/examples#17] [Sliced Review]
**Repository:** https://github.com/Affiction/angular-feature-sliced

**Methodology version:** 2.0

**Comment by azinit** [comment]
## 🟩 Начну с хорошего

[`praise:`](https://conventionalcomments.org/) Радует, что в целом структура соответствует некой конвенциональности с FSD, хоть и с небольшими расхождениями

![image](https://user-images.githubusercontent.com/42924400/150917774-638b8860-379e-4255-a263-27592f763ed8.png)

![image](https://user-images.githubusercontent.com/42924400/150917782-edddab15-d7a3-457d-a89d-0500cbce63a2.png)

![image](https://user-images.githubusercontent.com/42924400/150917791-b1364998-1e0a-4585-980b-243b7f76bc80.png)

**Reply by azinit** [reply]
[`nitpick(to-improve):`](https://conventionalcomments.org/) Как улучшить - можно все же попытаться придерживаться более конвенционального нейминга, чтобы не было потом конфликтов с [тулкитом FSD](https://www.npmjs.com/package/@feature-sliced/eslint-config)

- `components` => `ui`

[`nitpick(to-improve, public-api):`](https://conventionalcomments.org/) Ну и файлы рекэспорты внутри модулей тоже бы не помешали

**Reply by azinit** [reply]
Т.е. прям значительно проще становится, открывая такую структуру
- хоть assets и не держал бы [так глобально](https://github.com/feature-sliced/documentation/discussions/397#discussioncomment-1972610)
- и environments спрятать ближе к `shared/config` (но тут возможно уже специфика ангуляра, точно не знаю)

![image](https://user-images.githubusercontent.com/42924400/150918409-d6148e79-ecf7-4f46-8fad-6d7d4d5284f5.png)

**Reply by Affiction** [reply]
> * хоть assets и не держал бы [так глобально](https://github.com/feature-sliced/documentation/discussions/397#discussioncomment-1972610)
> * и environments спрятать ближе к `shared/config` (но тут возможно уже специфика ангуляра, точно не знаю)

Это некий дефолт в проекте, не хотелось бы трогать такие вещи. 
assets и environments(альтернатива shared/config) на любом чистом проекте уже законфигурены в angular.json.

**Reply by azinit** [reply]
Так и понял)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Тоже верно спроектировано, что store инициализируется в app

![image](https://user-images.githubusercontent.com/42924400/150918027-4c6a240a-34c8-44b5-9db0-4a504bb9c0e7.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) То что логика роутинга находится в одном месте (да и еще близко к pages) - тоже крутецки!

![image](https://user-images.githubusercontent.com/42924400/150918103-a6977c72-7a3c-48ca-91d2-2ab98e3ef64e.png)

![image](https://user-images.githubusercontent.com/42924400/150918111-8913034b-4957-4a0e-a356-f9d18738e117.png)

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Страницы тоже сделаны по большей части по канону:
- Композируется логика из нижележащих слоев
- При этом логика страницы остается максимально тонкой


![image](https://user-images.githubusercontent.com/42924400/150918272-dc8a0115-ed51-437e-bb30-16dc706e859b.png)

![image](https://user-images.githubusercontent.com/42924400/150918288-f717bd42-d935-4ac6-a4ae-36feb1b3ef08.png)

**Reply by azinit** [reply]
[`praise:`](https://conventionalcomments.org/) Тоже здорово, что удается грамотно композировать локальные сегменты и нижележащие слои

![image](https://user-images.githubusercontent.com/42924400/150918346-044dab40-1767-4688-bbe5-a102aa776369.png)

**Reply by Lonli-Lokli** [reply]
Разве импорты на 2 уровня вверх хорошо? ConfigService

**Reply by azinit** [reply]
@Lonli-Lokli 
https://github.com/feature-sliced/documentation/discussions/52#discussioncomment-477443

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/) Тоже канонично получается, что вся доменная логика ближе к entities расположена

![image](https://user-images.githubusercontent.com/42924400/150918590-16ecee89-1657-407e-a856-d4782abca849.png)

**Comment by azinit** [comment]
## 🟥 Теперь криты)

[`issue(public-api):`](https://conventionalcomments.org/) Очень много мест, где внутренности модулей **импортируются напрямую, минуя уровень public-api**

[`suggestion(public-api):`](https://conventionalcomments.org/) Лучше в указанных местах объявить файлы-реэкспорты, хоть у ангуляра и есть для этого отдельно "**.module.ts" подход. Хотя бы на уровне *слайсов*
(хотя можешь тут как раз таки развернуть - почему возможно PublicAPI с реэкспортом тут будет избыточен тогда)

![image](https://user-images.githubusercontent.com/42924400/150918678-c23eebe2-5006-470b-8dd9-4ac3d3d72654.png)

**Reply by azinit** [reply]
P.S. Получить такие ворнинги можно через [eslint-config](https://www.npmjs.com/package/@feature-sliced/eslint-config) 😏

**Reply by azinit** [reply]
Вот как частные примеры нарушения Public API

![image](https://user-images.githubusercontent.com/42924400/150918987-546b473a-1a8e-4488-9900-f73be56914d4.png)

![image](https://user-images.githubusercontent.com/42924400/150919002-3cacc4ee-342e-495f-add9-3ce59d64fc37.png)

**Comment by azinit** [comment]
[`issue(conventionality, naming):`](https://conventionalcomments.org/) Уже много раз [обсуждали](https://github.com/feature-sliced/documentation/discussions/418), касаемо проблемы конвенциональности

Да, безусловно, никто не может запретить использовать собственные названия слоев/сегментов
Но и надеяться, что это будет лучше сказываться на коммуникации и профитности тулкита - тоже не стоит (пока что) 🤔 

[`suggestion(conventionality, naming):`](https://conventionalcomments.org/) По возможности, как писал выше, лучше бы все же переименовать `components` в `ui`)

(либо же можешь описать, почему не зашел вариант с ui)



![image](https://user-images.githubusercontent.com/42924400/150919063-69be3535-35e2-43db-8fc3-b5b54ff3022e.png)

**Reply by Affiction** [reply]
Мне кажется что будет меньше конфузить ангуляр девов вариант `components/**.component.ts` чем `ui/**.component.ts`.
Обычно в компоненты именуются c суффиксом - RowComponent, файлы - row.component.(html|css|ts). Часто видел `feature-module/components/**/*.component.ts` и `feature-module/containers/**/*.container.ts`.

Я бы оставил на откуп разработчикам нейминг сегмента `ui`. А то в некоторых фреймворках устаканился определенный нейминг

**Reply by azinit** [reply]
А если надо будет положить куда-то логику, не относящуюся к компонентам, но так же реализующими UI-логику? 🤔 
(тот же canvas, или всякие mapbox-like решения? куда их кладете в ангуляре?)

**Reply by Affiction** [reply]
Надеюсь понял правильно)

Чтобы упростить компоненту часть логики можно выносить в сервис/презентер/директиву и положить рядом с основным компонентом
```
my-page/
  components/
    foo/
      baz.directive.ts  # если связанность высокая компоненты и директивы. Случай крайне редкий.
      foo.presenter.ts
      foo.service.ts
      foo.component.ts
```

В остальном всё нормально можно натянуть на FSD, что-то реюзабельное идёт в `lib`, БЛ в model

`pages/my-page/lib/foo.directive.ts` 
`pages/my-page/model/foo.service.ts`
`shared/lib/directives/foo.directive.ts` etc.

**Reply by azinit** [reply]
Ну т.е. angular в этом плане прям `component-oriented`, и все крутится вокруг него, если мы говорим про реализацию UI-логики, да?

**Reply by Affiction** [reply]
Всё верно, все те же компоненты, только добавили дополнительно парочку абстракций чтобы не придумывали велосипеды)

**Reply by azinit** [reply]
Понял, спасиб за разъяснения)

@feature-sliced/core будут мысли какие по ситуации?

**Comment by azinit** [comment]
[`praise:`](https://conventionalcomments.org/#labels) Еще касаемо api - тоже получается довольно канонично)

- И разделение по сервисам, как внешних источниках данных
- И изоляция логики конкретного ресурса - в самой директории ресурса
- И общая декларация внешнего интерфейса для использования api

![image](https://user-images.githubusercontent.com/42924400/150919721-2c8f6386-9c7a-4db0-aac7-efcd826e4607.png)

![image](https://user-images.githubusercontent.com/42924400/150919700-d9f70ca4-9ea5-4d96-97a1-de882547862a.png)

**Comment by azinit** [comment]
[`praise(routes):`](https://conventionalcomments.org/#labels) Еще из хорошего - что в shared тащишь только общую логику по работе с роутами, но не саму специфику роутов (от этого [тоже есть проблемы](https://github.com/feature-sliced/examples/discussions/14#discussioncomment-1972571))

![image](https://user-images.githubusercontent.com/42924400/150919837-f54ed9f6-0dc6-4755-9d8e-9fba62dc9f6a.png)

**Reply by azinit** [reply]
[`nitpick:`](https://conventionalcomments.org/) Разве что - public api тоже бы унифицировать

[`thoughts:`](https://conventionalcomments.org/)  А то получается, что можно использовать как `routerModel.RouterFacade` , так и `RouterFacade` отдельно

Может потом проблем при рефакторинге создать 🤔 
![image](https://user-images.githubusercontent.com/42924400/150920065-30fae53d-3503-414f-a4e3-213a94c81418.png)

**Reply by azinit** [reply]
[`nitpick(to-improve, non-blocking):`](https://conventionalcomments.org/) Тут кста, та же проблема с public api)

![image](https://user-images.githubusercontent.com/42924400/150920401-9e405184-ba48-4cfa-a13d-73fb99cf585e.png)

**Reply by Affiction** [reply]
> [`nitpick:`](https://conventionalcomments.org/) Разве что - public api тоже бы унифицировать
> 
> [`thoughts:`](https://conventionalcomments.org/) А то получается, что можно использовать как `routerModel.RouterFacade` , так и `RouterFacade` отдельно
> 
> Может потом проблем при рефакторинге создать 🤔

RouterFacade сервис используется чисто для DI. 

Хотелосьбы писать:
```typescript
class FooComponent {
    constructor(private routerFacade: RouterFacade){}
}
```
вместо:
```typescript
class FooComponent {
    constructor(private routerFacade: routerModel.RouterFacade){}
}
```

Потому экспортируется и то, и то, в противном случае выйдет что мы нарушим правило с публичным апи
```typescript
export * as TaskModel from './model'; // экспортируем всё кроме TaskFacade
export * as TaskLib from './lib';
export { TaskFacade } from './model/task.facade.ts'; // проблема появляется здесь
export { TaskModule } from './task.module';
```

Есть идеи как разрулить такое?)

**Reply by azinit** [reply]
Да ощущение, что все это резолвится банальным реэкспортов юнитов из сегментов таким образом, чтобы при использовании не надо было лезть во внутренности (т.е. как ты описал как раз)

Другой вопрос, насколько стоит так усложнять PublicAPI, чтобы до одних и тех же юнитов можно было разными путями доходить 🤔 
Потом при рефакторинге наверняка выстрелить может)

> Но с DI в целом непросто может быть, так что понимаю
(хотя в целом пока не особо вижу проблемы использования `routerModel.RouterFacade`)

**Comment by azinit** [comment]
## 🟨 Теперь non-blocking 

> Осталось немного 😏 

[`nitpick(to-improve, non-blocking):`](https://conventionalcomments.org/) Некрит, но можно попробовать придерживаться порядка импортов для слоев - для консистентности

> Подробнее [тут](https://github.com/feature-sliced/eslint-config/tree/master/rules/import-order), автофиксится спокойно

![image](https://user-images.githubusercontent.com/42924400/150920233-656ed438-31f9-4a2d-8234-a3c891ef21f4.png)

**Comment by azinit** [comment]
[`nitpick(to-improve, non-blocking, public-api):`](https://conventionalcomments.org/)  Некрит, но для app и environments - тоже бы файлы реэкспорты объявить

![image](https://user-images.githubusercontent.com/42924400/150920486-3215f3dd-4ba8-43c3-8bf5-e7a40b0f88f9.png)

![image](https://user-images.githubusercontent.com/42924400/150920495-256a49a4-5df6-4362-8a2b-c5a9cdcf8005.png)

**Comment by azinit** [comment]
[`nitpick(to-improve, non-blocking, naming):`](https://conventionalcomments.org/) Некрит, но если есть возможность не дублировать неймспейсы в структуре - будет лучше
(и для рефакторинга, и для восприятия)

Т.е. тут `toggle-task/**/toggle-task` излишен, как если бы было `pages/some/some-page-content/...`

![image](https://user-images.githubusercontent.com/42924400/150920560-f1a388af-039a-4015-8e58-423e2c617283.png)

**Reply by Affiction** [reply]
Обычно компоненты генерируются через cli. На выходе получаем `foo.component.(html|css|ts)`, внутри `FooComponent` и в метаданных указан `selector: 'prefix-foo'`. С одной стороны можно изначально сгенерировать просто card, но тогда внутри нужно переименовать selector компонента на уникальный, если 2 одинаковых селектора будет, то компилятор выдаст ошибку. Если честно, нигде не видел чтобы такое практиковали с разным неймингом селектора и файлов)

**Reply by azinit** [reply]
Эхх, опять специфика ангуляра ставит палки)
Но оке, если это конвенционально, то тут ничего не сможем сделать
Просто я в целом к тому, чтобы не порождать модули типа `pages/some/some-page-content`, т.к. это избыточный нейминг)

@feature-sliced/core не дадут соврать, уже думаем этот антипаттерн в доку занести))

**Comment by azinit** [comment]
[`nitpick(to-improve, non-blocking, shared):`](https://conventionalcomments.org/) Некрит, но точно ли нужно для shared uikit отдельный сегмент создавать, если идет подвязка на внутреннюю либу? 🤔 

![image](https://user-images.githubusercontent.com/42924400/150920678-2a02d67d-477c-4e54-b331-a79efa1f8de3.png)

**Reply by azinit** [reply]
[`nitpick(to-improve, non-blocking, shared):`](https://conventionalcomments.org/)  Ну и тут тоже - есть сомнения определенные, что к любому сегменту shared стоит обращаться через единый файл реэкспорт
(почти то же, что для entities и features делать такой же общий `Layer PublicAPI` 😏 )

![image](https://user-images.githubusercontent.com/42924400/150920835-33e96715-d791-4b7d-ae23-62ad796377ab.png)

Особенно учитывая - что там только UIKit, даже без lib

![image](https://user-images.githubusercontent.com/42924400/150920964-9bd3f8b5-160b-454d-9a95-24926fd4e9a9.png)

**Reply by Affiction** [reply]
Здесь не знал куда положить)

`UiKitModule` закинул в `share/ui`
`SharedModule` закинул в `shared/lib`

**Comment by azinit** [comment]
[`nitpick(to-improve, non-blocking, model):`](https://conventionalcomments.org/) Некрит и внутреннее устройство model - зависит от команды и стека

Но выделять на каждый тип юнита отдельный файл - достаточно проблемно может быть при масштабировании
(эффектор уже от этого отказался, а redux пока увы продолжает так практиковать)

Возможно у rxjs такие же практики - тогда ок)
Но если есть возможность держать это все добро в одном файле - то мб и к лучшему? 🤔 

![image](https://user-images.githubusercontent.com/42924400/150920986-d7a71aac-35c8-48f3-9a14-be651be7d7c7.png)

**Reply by azinit** [reply]
Хотя это даже не `non-blocking` , а скорее размышления

**Reply by Affiction** [reply]
Сторы генерируются схематиком nx. Структура здесь +- дефолтная. В store-name.models.ts обычно лежит не конкретная моделька, а всё что относится к данному стору

**Comment by azinit** [comment]
## 🔳 Теперь чисто мысли

[`thoughts(structure):`](https://conventionalcomments.org/) Структура каждого модуля выглядит прям оверхедно, ввиду специфики ангуляра
А если еще добавить `**.module.ts` файлы - то вообще жесть)

Надеюсь тут или выше раскроешь подробней - почему это ОК для ангуляра


![image](https://user-images.githubusercontent.com/42924400/150921365-cba6a3c6-3b04-46ae-b4da-716164eaf1a2.png)

**Reply by azinit** [reply]
[`thoughts(public-api):`](https://conventionalcomments.org/) Тоже интересно, что приходится заранее объявлять импортируемые зависимости и экспортируемые юниты 

Но чет даже при всей строгости FSD кажется оверхедно, если такое пытаться "подружить" с файлами реэкспортами 🤔 

![image](https://user-images.githubusercontent.com/42924400/150921494-75ca76d1-3d47-4b62-b10a-675827072537.png)

**Reply by Affiction** [reply]
В Ангуляре всё построено на концепции модулей. Каждый модуль может быть вполне самостоятельной либкой. Потому с нашей стороны мы готовим модуль, импортируем зависимости которые будем использовать, например в компонентах, и экспортируем то с чем конечный пользователь может взаимодействовать. 

- В `declarations` мы указываем какие компоненты, директивы, пайпы относятся к этому модулю (нужно компилятору)
- В `imports` мы указываем какие сторонние модули нужны чтобы компоненты адекватно работали. Пример, в SharedModule экспортируется CommonModule, который позволяет в разметке использовать базовые дериктивы *ngIf, *ngFor, пайпы DatePipe и тд. Если не передать, то компилятор упадёт с ошибкой если будем пытаться использовать выше перечисленные директивы.
- В `exports` передаёт то что можно будет использовать в других модулях при импорте конкретного модуля. Так сказать некое публичное апи
- В `providers` регестрируются сущности для DI

Я бы не сказал что здесь будут какие-то проблемы. Пример, делаем фичу, сразу создаём ангуляр модуль под фичу, в ней регистрируем компонент в `declarations`, и добавляем в `exports` чтобы могли использовать в сторонних модулях. Дальше чтобы использовать на конкретной пейдже, мы добавляем модуль фичи в `imports` нашей пэйджи, и после этого можем спокойно использовать фичу в разметке, вставив простой тег `<prefix-my-feature></prefix-my-feature>`

**Reply by azinit** [reply]
Реактовцы рвут волосы на себе от такой "осознанной декларативности"))

Но в целом да, понимаю что специфика ангуляра во многом

**Reply by azinit** [reply]
Тут проблемно именно, что PublicAPI в случая ангуляра оформляется на двух уровнях:
- Файлы реэкспорты (чтобы во внутренности не лезть)
- **.module файлы (чтобы "компилятор ангуляра" подхватил)

**Comment by azinit** [comment]
## FINAL

Фух, постарался сильно не докапываться, но основные моменты расписал)

@Affiction, если получится хотя бы криты поправить - то уже можем заливать твой пример в доку

**Reply by azinit** [reply]
Ну и это тоже тут оставлю, чтоб не потерялось

https://www.npmjs.com/package/@feature-sliced/eslint-config

**Comment by Lonli-Lokli** [comment]
Появилась новая фишка - standalone components, компоненты без NgModule (как условный докер)

**Reply by azinit** [reply]
@Affiction FYI

**Reply by Affiction** [reply]
https://angular.io/guide/standalone-components
> The standalone component feature is available for developer preview. It's ready for you to try; but it might change before it is stable.


Пока не вижу чтобы фича глобально повлияла на структуру.
В будущем если фича приживётся, подрефачу продж

**Reply by Lonli-Lokli** [reply]
@Affiction А почему вообще не используется libs из Nx?

**Reply by Affiction** [reply]
@Lonli-Lokli на сколько помню, там нету никаких вещей которые могли бы стать либами
